### DISCUSSION TITLE
Plugin DynamicModelChoiceField Filtering

### DISCUSSION THREAD

**Discussion by cslingerland** [netbox-community/netbox#9929] [Q&A] [answered]
As I understand it, the DynamicModelChoiceField ties into the API to get results and build the list, but I'm not able to get this to work for the models in my Plugin. I have the base API setup and working but when I use this field class, I get no results. I have a workaround of using django.forms.ModelChoiceField, but it'd be nice if I could get DynamicModelChoiceField working.

Interestingly, when I edit an existing Relationship, the form does show the Vendor correctly, but it just has the one listed. It doesn't fetch any other results so I can't change the vendor.

Here's my form class:
```
class RelationshipForm(NetBoxModelForm):
    vendor = DynamicModelChoiceField(
        queryset=Vendor.objects.all(),
        required=False,
    )
    
    class Meta:
        model = Relationship
        fields = ('vendor', 'dc', 'notes', 'tags',)
```

**Comment by DanSheps** [comment, marked as answer]
If you open up your dev-tools after loading the page, what does it try to connect to after you open the choice field?  There should be a API query, that might give you some insight.

It populates correctly when it is  edited with an existing one because the data is pre-populated with that vendor.

**Reply by cslingerland** [reply]
I see what you mean now. It's in my browser console. I do have an error message there:
"Data object option must have at least have a text value. Check object: " and then it shows the dictionary of the Vendor object it's trying to call.

**Reply by cslingerland** [reply]
Full message: 
```
Data object option must have at least have a text value. Check object:
{"value":"67","data":{"name":"test","ports":"test","dc-presence":"Both","created":"2022-07-22T13:42:21.661490Z","last-updated":"2022-07-22T13:42:21.730978Z"}

```

**Reply by DanSheps** [reply]
You likely need a "display" property on your serializer:

```python
class VendorSerializer(NetBoxModelSerializer):
    ...
    class Meta:
        model = Vendor
        fields = [...'display'...]
```

I believe this maps to the __str__ method on the model so make sure that is populated as well.

**Reply by cslingerland** [reply]
That was it! I would not have figured that out so THANK YOU!

**Reply by Aldi0091** [reply]
thank you for raising and answering the issue, I had spent hours on it =) here just let it sorted out in a bit of second

**Reply by venkatvamshi** [reply]
Hi , 
I am seeing a similar issue with my SecretRoleSerializer(ValidatedModelSerializer).  When trying to filter using SecretRole. 

```
class SecretRoleSerializer(ValidatedModelSerializer):
    url = serializers.HyperlinkedIdentityField(view_name='*********')
    secret_count = serializers.IntegerField(read_only=True)

    class Meta:
        model = SecretRole
        fields = ['id', 'url', 'name', 'slug', 'description', 'secret_count']
```
and my secretRole model has the `str` method mapping to `name` which is included in the fields. 

Can someone help me with this please

<img width="1234" alt="logs" src="https://user-images.githubusercontent.com/10359735/233418225-0d3d29aa-a545-449c-8e64-ba052707104b.png">

**Reply by abhi1693** [reply]
You are missing `display` field. Check https://github.com/Onemind-Services-LLC/netbox-secrets/blob/master/netbox_secrets/api/serializers.py#L58

**Reply by venkatvamshi** [reply]
Thanks, This worked.
